pkg/controllers/management/clusterprovisioner/provisioner.go

```diff
@@ -214,6 +245,39 @@ func (p *Provisioner) Updated(cluster *v3.Cluster) (runtime.Object, error) {
 	return obj.(*v3.Cluster), err
 }
 
+func (p *Provisioner) startCleanUpHelmReleases(cluster *v3.Cluster) error {
+	err := p.cleanupHelmReleases(cluster)
+	if err != nil {
+		return err
+	}
+
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		cluster, err = p.Clusters.Get(cluster.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// Attempt to manually trigger updating, otherwise it will not be triggered until after exiting reconcile
+		apimgmtv3.ClusterConditionHelmReleasesMigrated.True(cluster)
+		apimgmtv3.ClusterConditionHelmReleasesMigrated.Message(cluster, "Helm releases cleaned up")
+
+		cluster, err = p.Clusters.Update(cluster)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// shouldCleanHelmReleases indicates whether the cluster upgrade process should clean up existing Helm releases
+func shouldCleanHelmReleases(helmCleanupOnUpgradeEnabled bool, cluster *v3.Cluster) bool {
+	return helmCleanupOnUpgradeEnabled &&
+		!apimgmtv3.ClusterConditionHelmReleasesMigrated.IsTrue(cluster) &&
```

Comment:
nit: it doesn't make a huge different but in case you didn't know conditions have a function called `DoUntilTrue` where anything you wrap in it will only be done if the condition isn't true. It will not retry until true like the name might imply.

Reply:
How would the code look like if I were to use `DoUntilTrue`?